src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java

```diff
@@ -0,0 +1,432 @@
+            List<Header> combinedHeaders = NetworkUtility.combineHeaders(responseHeaders, entry);
+            callback.onSuccess(
+                    new NetworkResponse(
+                            HttpURLConnection.HTTP_NOT_MODIFIED,
+                            entry.data,
+                            /* notModified= */ true,
+                            SystemClock.elapsedRealtime() - requestStartMs,
+                            combinedHeaders));
+            return;
+        }
+
+        byte[] responseContents = httpResponse.getContentBytes();
+
+        if (responseContents == null) {
+            final InputStream inputStream = httpResponse.getContent();
+            if (inputStream == null) {
+                // Add 0 byte response as a way of honestly representing a
+                // no-content request.
+                responseContents = new byte[0];
+            } else {
+                Runnable run =
+                        new Runnable() {
+                            @Override
+                            public void run() {
+                                PoolingByteArrayOutputStream bytes =
+                                        new PoolingByteArrayOutputStream(
+                                                mPool, httpResponse.getContentLength());
+                                byte[] buffer = new byte[1024];
+                                int count = 0;
+                                while (true) {
+                                    try {
+                                        if ((count = inputStream.read(buffer)) == -1) break;
+                                    } catch (IOException e) {
+                                        onRequestFailed(
+                                                request,
+                                                callback,
+                                                e,
+                                                requestStartMs,
+                                                httpResponse,
+                                                bytes.toByteArray());
+                                    }
+                                    bytes.write(buffer, 0, count);
+                                }
+                                byte[] finalResponseContents = bytes.toByteArray();
+                                runAfterBytesReceived(
+                                        requestStartMs,
+                                        statusCode,
+                                        httpResponse,
+                                        request,
+                                        callback,
+                                        responseHeaders,
+                                        finalResponseContents);
+                            }
+                        };
+                mBlockingExecutor.execute(run);
+                return;
+            }
+        }
+        runAfterBytesReceived(
```

Comment:
nit: since this only applies to the case that the response bytes are already read, which is checked many lines above, I'd put this up there and short-circuit return in that case, and then do the rest of the handling for the stream reading case.